packages/apps/plugins/plugin-graph/src/graph/graph.ts

```diff
@@ -0,0 +1,189 @@
+    if (!path) {
+      return undefined;
+    }
+
+    return path.length > 0 ? get(this._root, path) : this._root;
+  }
+
+  traverse({ from = this._root, direction = 'down', predicate, onVisitNode }: Graph.TraverseOptions): void {
+    if (!predicate || predicate(from)) {
+      onVisitNode?.(from);
+    }
+
+    if (direction === 'down') {
+      Object.values(from.children).forEach((child) => this.traverse({ from: child, predicate, onVisitNode }));
+    } else if (direction === 'up' && from.parent) {
+      this.traverse({ from: from.parent, direction, predicate, onVisitNode });
+    }
+  }
+
+  registerNodeBuilder(builder: Graph.NodeBuilder): void {
+    this._nodeBuilders.add(builder);
+  }
+
+  removeNodeBuilder(builder: Graph.NodeBuilder): void {
+    this._nodeBuilders.delete(builder);
+  }
+
+  construct(from = this._root as Graph.Node, path: string[] = [], ignoreBuilders: Graph.NodeBuilder[] = []): void {
+    this._index[from.id] = path;
+    const subscriptions = this._unsubscribe.get(from.id) ?? new EventSubscriptions();
+    subscriptions.clear();
+    Array.from(this._nodeBuilders.values())
+      .filter((builder) => ignoreBuilders.findIndex((ignore) => ignore === builder) === -1)
+      .forEach((builder) => {
+        const unsubscribe = builder(this._filterBuilders(from, path, [...ignoreBuilders, builder]));
+        unsubscribe && subscriptions.add(unsubscribe);
+      });
+    this._unsubscribe.set(from.id, subscriptions);
+  }
+
+  // Updates a Node's add method to filter out builders that have already been applied.
+  private _filterBuilders(node: Graph.Node, path: string[], ignoreBuilders: Graph.NodeBuilder[]): Graph.Node {
+    const builderNode = deepSignal({
+      ...node,
+      add: (...partials) => {
+        return partials.map((partial) => {
+          const childPath = [...path, 'childrenMap', partial.id];
+          const child = this._createNode({ ...partial, parent: builderNode }, childPath, ignoreBuilders);
+          builderNode.childrenMap[child.id] = child;
+          this.construct(child, childPath, ignoreBuilders);
+          return child;
+        });
+      },
+    }) as RevertDeepSignal<Graph.Node>;
+
+    return node;
+  }
+
+  private _createNode<TData = null, TProperties extends { [key: string]: any } = { [key: string]: any }>(
+    partial: Pick<Graph.Node, 'id' | 'label'> & Partial<Graph.Node<TData, TProperties>>,
+    path: string[] = [],
+    ignoreBuilders: Graph.NodeBuilder[] = [],
+  ): Graph.Node<TData, TProperties> {
+    const node: Graph.Node<TData, TProperties> = deepSignal({
+      data: null as TData,
+      parent: null,
+      properties: {} as TProperties,
+      childrenMap: {},
+      actionsMap: {},
+      ...partial,
+      // TODO(wittjosiah): Default sort.
+      get children() {
+        return Object.values(node.childrenMap);
+      },
+      get actions() {
+        return Object.values(node.actionsMap);
+      },
+      add: (...partials) => {
```

Comment:
Since Node isn't pure (has functions) then make it an actual class rather than a struct with functions attached?
Is this a deepsignal concern?

Reply:
Yeah, I wanted to take advantage of deepsignal turns getters into `computed` calls in signal. I'm not sure if that works within classes, it might? I haven't tried, but also having deepsignal venture down into classes is often when I've had the most trouble with it, so I've just generally tried to avoid that (which is part of why it's used _inside_ this class rather than dropping this class into a deepsignal).

Reply:
Actually there are plenty of reasons not to apply classes without good rationale. They are now a rather weathered feature, and there's plenty of criticism of the `class` keyword and the notorious `this` keyword out there, and while I don't consider them to be serious drawbacks I think it's enough to warrant some thought about whether we're applying classes because inheritance or some other "valuable" property of classes is what we need here (and it isn't better solved with something like "composition" etc). fwiw if it aint broke we dont need to fix it, and writing expressions inline can be helpful to both read and write code and build static analysis tooling against (something we'll actually likely need to do soon in the type-safety of ECHO stream). It's also less verbose than writing classes. Either way the space for problems where OOP patterns are required is shrinking.

https://github.com/getify/You-Dont-Know-JS/blob/1st-ed/this%20&%20object%20prototypes/README.md#you-dont-know-js-this--object-prototypes
https://en.wikipedia.org/wiki/Composition_over_inheritance